platform/gtk4/cog-platform-gtk4.c

```diff
@@ -0,0 +1,749 @@
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+
+    glActiveTexture(GL_TEXTURE0);
+    glUniform1i(window->gl_texture_location, 0);
+
+    /* Configuration for the FDO backend. */
+    EGLDisplay display = eglGetCurrentDisplay();
+    g_assert(display != EGL_NO_DISPLAY);
+
+    EGLint egl_major = 0, egl_minor = 0;
+    if (eglInitialize(display, &egl_major, &egl_minor) == EGL_FALSE) {
+        g_set_error_literal(error, COG_PLATFORM_EGL_ERROR, eglGetError(),
+            "Cannot initialize EGL");
+        return false;
+    }
+
+    g_debug("EGL %i.%i successfully initialized.", egl_major, egl_minor);
+    wpe_fdo_initialize_for_egl_display(display);
+    return true;
+}
+
+static void
+realize(GtkWidget* widget, gpointer user_data)
+{
+    struct platform_window* win = user_data;
+    g_autoptr(GError) error = NULL;
+    if (!setup_shader(win, &error)) {
+        g_warning("Shader setup failed: %s", error->message);
+        g_application_quit(g_application_get_default());
+    }
+}
+
+static gboolean
+render(GtkGLArea* area, GdkGLContext* context, gpointer user_data)
+{
+    struct platform_window* win = user_data;
+    glClearColor(0, 0, 0, 0);
+    glClear(GL_COLOR_BUFFER_BIT);
+    glViewport(0, 0, win->width, win->height);
+
+    if (win->commited_image)
+        wpe_view_backend_exportable_fdo_egl_dispatch_release_exported_image(
+            win->exportable, win->commited_image);
+
+    if (!win->current_image) {
+        gtk_gl_area_queue_render(GTK_GL_AREA(win->gl_drawing_area));
+        return TRUE;
+    }
+
+    glUseProgram(win->gl_program);
+
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(GL_TEXTURE_2D, win->gl_texture);
+    win->ext_glEGLImageTargetTexture2DOES(
+        GL_TEXTURE_2D,
+        wpe_fdo_egl_exported_image_get_egl_image(win->current_image));
+    glUniform1i(win->gl_texture_location, 0);
+
+    win->commited_image = win->current_image;
+
+    static const GLfloat vertices[4][2] = {
+        { -1.0, 1.0 },
+        { 1.0, 1.0 },
+        { -1.0, -1.0 },
+        { 1.0, -1.0 },
+    };
+
+    static const GLfloat texturePos[4][2] = {
+        { 0, 0 },
+        { 1, 0 },
+        { 0, 1 },
+        { 1, 1 },
+    };
+
+    glVertexAttribPointer(0, 2, GL_FLOAT, GL_FALSE, 0, vertices);
+    glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 0, texturePos);
+
+    glEnableVertexAttribArray(0);
+    glEnableVertexAttribArray(1);
+
+    glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+
+    glDisableVertexAttribArray(0);
+    glDisableVertexAttribArray(1);
+
+    glBindTexture(GL_TEXTURE_2D, 0);
+
+    wpe_view_backend_exportable_fdo_dispatch_frame_complete(win->exportable);
+    return TRUE;
+}
+
+void resize(GtkGLArea* area, int width, int height, gpointer user_data)
```

Comment:
Let's make this `static`. Also the return type and function attributes should go in their own line:

```suggestion
static void
resize(GtkGLArea* area, int width, int height, gpointer user_data)
```

(The same goes for all the other functions around which have the return type in the same line as the name and parameters.)